components/brave_wallet/browser/zcash/zcash_orchard_storage_unittest.cc

```diff
@@ -0,0 +1,345 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/zcash/zcash_orchard_storage.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/files/scoped_temp_dir.h"
+#include "brave/components/brave_wallet/browser/zcash/zcash_test_utils.h"
+#include "brave/components/brave_wallet/common/common_utils.h"
+#include "content/public/test/browser_task_environment.h"
+#include "testing/gtest/include/gtest/gtest.h"
+
+namespace brave_wallet {
+
+class OrchardStorageTest : public testing::Test {
+ public:
+  OrchardStorageTest()
+      : task_environment_(base::test::TaskEnvironment::TimeSource::MOCK_TIME) {}
+  void SetUp() override;
+
+  base::test::TaskEnvironment task_environment_;
+  base::ScopedTempDir temp_dir_;
+  std::unique_ptr<ZCashOrchardStorage> orchard_storage_;
+};
+
+void OrchardStorageTest::SetUp() {
+  ASSERT_TRUE(temp_dir_.CreateUniqueTempDir());
+  base::FilePath db_path(
+      temp_dir_.GetPath().Append(FILE_PATH_LITERAL("orchard.db")));
+  orchard_storage_ = std::make_unique<ZCashOrchardStorage>(db_path);
+}
+
+TEST_F(OrchardStorageTest, AccountMeta) {
+  auto account_id_1 = MakeIndexBasedAccountId(mojom::CoinType::ZEC,
+                                              mojom::KeyringId::kZCashMainnet,
+                                              mojom::AccountKind::kDerived, 0);
+  auto account_id_2 = MakeIndexBasedAccountId(mojom::CoinType::ZEC,
+                                              mojom::KeyringId::kZCashMainnet,
+                                              mojom::AccountKind::kDerived, 1);
+
+  {
+    auto result = orchard_storage_->GetAccountMeta(account_id_1.Clone());
+    EXPECT_FALSE(result.has_value());
+  }
+
+  EXPECT_FALSE(
+      orchard_storage_->RegisterAccount(account_id_1.Clone(), 100, "hash"));
+
+  {
+    auto result = orchard_storage_->GetAccountMeta(account_id_1.Clone());
+    EXPECT_TRUE(result.has_value());
+    EXPECT_EQ(result->account_birthday, 100u);
+    EXPECT_EQ(result->latest_scanned_block_id, 100u);
+    EXPECT_EQ(result->latest_scanned_block_hash, "hash");
+  }
+
+  // {
+  //   // Failed to insert same account
+  //   EXPECT_EQ(orchard_storage_->RegisterAccount(account_id_1.Clone(), 200,
+  //   "hash").value().error_code,
+  //             OrchardStorage::OrchardStorageErrorCode::kFailedToExecuteStatement);
+  // }
+
+  // Insert second account
+  EXPECT_FALSE(
+      orchard_storage_->RegisterAccount(account_id_2.Clone(), 200, "hash"));
+  {
+    auto result = orchard_storage_->GetAccountMeta(account_id_2.Clone());
+    EXPECT_TRUE(result.has_value());
+    EXPECT_EQ(result->account_birthday, 200u);
+    EXPECT_EQ(result->latest_scanned_block_id, 200u);
+    EXPECT_EQ(result->latest_scanned_block_hash, "hash");
+  }
+}
+
+TEST_F(OrchardStorageTest, PutDiscoveredNotes) {
```

Comment:
What happens if we spend a note from the wrong account? E.g. Account 1 spends note 111. It would be good to have some error tests for these types of scenarios.

Reply:
We don't have functionality to spend notes yet

Reply:
Gotcha, ok I'll check for a test for that once we've added it. On second look, makes sense that right now we're just setting this DB up and could do whatever we want with it so it would be too low of a layer to test for this.